osu.Game.Rulesets.Catch/UI/HitExplosion.cs

```diff
@@ -1,129 +1,56 @@
-                    RelativeSizeAxes = Axes.Both,
-                    Masking = true,
-                    Size = new Vector2(0.01f, initial_height),
-                    Blending = BlendingParameters.Additive,
-                },
-                directionalGlow2 = new CircularContainer
-                {
-                    Anchor = Anchor.Centre,
-                    Origin = Anchor.Centre,
-                    RelativeSizeAxes = Axes.Both,
-                    Masking = true,
-                    Size = new Vector2(0.01f, initial_height),
-                    Blending = BlendingParameters.Additive,
-                }
+                CentreComponent = false,
+                Anchor = Anchor.BottomCentre,
+                Origin = Anchor.BottomCentre
             };
         }
 
         protected override void OnApply(HitExplosionEntry entry)
         {
-            X = entry.Position;
-            Scale = new Vector2(entry.Scale);
-            setColour(entry.ObjectColour);
-
-            using (BeginAbsoluteSequence(entry.LifetimeStart))
-                applyTransforms(entry.RNGSeed);
+            base.OnApply(entry);
+            if (IsLoaded)
+                apply(entry);
         }
 
-        private void applyTransforms(int randomSeed)
+        protected override void LoadComplete()
         {
-            ClearTransforms(true);
-
-            const double duration = 400;
-
-            // we want our size to be very small so the glow dominates it.
-            largeFaint.Size = new Vector2(0.8f);
-            largeFaint
-                .ResizeTo(largeFaint.Size * new Vector2(5, 1), duration, Easing.OutQuint)
-                .FadeOut(duration * 2);
-
-            const float angle_variangle = 15; // should be less than 45
-            directionalGlow1.Rotation = StatelessRNG.NextSingle(-angle_variangle, angle_variangle, randomSeed, 4);
-            directionalGlow2.Rotation = StatelessRNG.NextSingle(-angle_variangle, angle_variangle, randomSeed, 5);
-
-            this.FadeInFromZero(50).Then().FadeOut(duration, Easing.Out).Expire();
+            base.LoadComplete();
+            apply(Entry);
         }
 
-        private void setColour(Color4 objectColour)
+        private void apply(HitExplosionEntry entry)
         {
-            const float roundness = 100;
-
-            largeFaint.EdgeEffect = new EdgeEffectParameters
-            {
-                Type = EdgeEffectType.Glow,
-                Colour = Interpolation.ValueAt(0.1f, objectColour, Color4.White, 0, 1).Opacity(0.3f),
-                Roundness = 160,
-                Radius = 200,
-            };
-
-            smallFaint.EdgeEffect = new EdgeEffectParameters
-            {
-                Type = EdgeEffectType.Glow,
-                Colour = Interpolation.ValueAt(0.6f, objectColour, Color4.White, 0, 1),
-                Roundness = 20,
-                Radius = 50,
-            };
+            ApplyTransformsAt(double.MinValue, true);
```

Comment:
This is not visible currently, but this `ApplyTransformsAt` is not doing what is intended because transforms of drawables are removed when completed by default. (It is not enough to override `RemoveCompletedTransforms` in `LegacyHitExplosion` because child `Sprite` transforms are removed).
I guess this recursive variant of `ApplyTransformsAt` is error prone.

Reply:
Yes, well, it's more for safety than anything (ie. to handle cases in which the explosion can die while animating, such as non-frame-stable seeks in editor).

Reply:
I don't understand how this line is is useful (adding safety) in such case. If a transform is cleared (in the next line) and the values transformed are restored (in `Animate`, not here as stated above) then nothing seems affected by the transforms that is canceled midway?

Reply:
I'll remove it to not prolong this conversation further. I feel safer with it in but it should indeed work either way.

Reply:
Actually... no, I take that above comment back. Removing the apply call actually broke stuff, so it must have done *something*.

And it does, because `PoolableDrawableWithLifetime` has `RemoveCompletedTransforms => false`, and `CompositeDrawable`s [actually propagate that flag to children on add](https://github.com/ppy/osu-framework/blob/f6fc0b49f74fd78241c038cb3ce60cbfcd584a74/osu.Framework/Graphics/Containers/CompositeDrawable.cs#L546) and [on set](https://github.com/ppy/osu-framework/blob/f6fc0b49f74fd78241c038cb3ce60cbfcd584a74/osu.Framework/Graphics/Containers/CompositeDrawable.cs#L1194-L1195). So in the draw visualiser *all* nested drawables also have `RemoveCompletedTransforms = false`.

Reply:
I'm going to resolve this convo and go with this as is, because other implementations (taiko's `HitExplosion` and `DrawableJudgement`) do the same thing.